src/stratigraphy/groundwater/utility.py

```diff
@@ -0,0 +1,126 @@
+"""Series of utility functions for groundwater stratigraphy."""
+
+import regex
+
+
+def extract_date(text: str) -> str | None:
+    """Extract the date from a string."""
+    date = None
+    date_match = regex.search(r"(\d{2}\.\d{2}\.\d{4})", text)
+    if date_match:
+        date = date_match.group(1)
+    else:
```

Comment:
I am not super happy about this construct. I think it became like this every time I added a new rule.

Reply:
How about something like

```
patterns = [
    r"(\d{2}\.\d{2}\.\d{4})", # some comment
    ...
]
for pattern in patterns:
    date_match = regex.search(pattern, text)
    if date_match:
        return date_match.group(1)
```

That would get rid of most of the code duplication and the excessively nested if-clauses.

Reply:
I like this idea. Thank you! I will implement that!

Reply:
Also, why not immediately match all cases at once, like `(\d{1,2}\.\d{1,2}\.\d{2,4})`?

Also, we need to make sure that we don't stop the rexeg in the middle of a number. Because the current implementation only extracts "9.1.20" from "9.1.2018".